File: frontend/src/api/audit.ts
import { get } from "./client";
import type { AuditListResponse, AuditRunDetail, AuditFilters } from "./contracts";

/** Fetch paginated audit run list from GET /api/audit */
export async function fetchAuditRuns(
  filters?: AuditFilters,
  signal?: AbortSignal,
): Promise<AuditListResponse> {
  const params: Record<string, unknown> = {};
  if (filters?.team)          params.team        = filters.team;
  if (filters?.environment)   params.environment = filters.environment;
  if (filters?.status)        params.status      = filters.status;
  if (filters?.artifact_type) params.artifact_type = filters.artifact_type;
  if (filters?.page)          params.page        = filters.page;
  if (filters?.limit)         params.limit       = filters.limit;

  return get<AuditListResponse>("/api/audit", params, signal);
}

/** Fetch a single run detail from GET /api/audit/:run_id */
export async function fetchRunDetail(
  run_id: string,
  signal?: AbortSignal,
): Promise<AuditRunDetail> {
  return get<AuditRunDetail>(`/api/audit/${run_id}`, undefined, signal);
}

File: backend/routes/audit.py
"""
audit.py
GET /api/audit             — paginated, filterable list of all validation runs
GET /api/audit/<run_id>    — single run detail
"""

from flask import Blueprint, request, jsonify

audit_bp = Blueprint("audit", __name__)


def _serialize(doc: dict) -> dict:
    """Convert MongoDB doc to JSON-safe dict."""
    doc.pop("_id", None)
    if "timestamp" in doc and hasattr(doc["timestamp"], "isoformat"):
        doc["timestamp"] = doc["timestamp"].isoformat()
    return doc


@audit_bp.route("/api/audit", methods=["GET"])
def get_audit_logs():
    from app import db
    if db is None:
        return jsonify({"runs": [], "total": 0, "page": 1, "limit": 20}), 200

    # ── Query params ───────────────────────────────────────────────────────────
    team        = request.args.get("team")
    environment = request.args.get("environment")
    status      = request.args.get("status")
    page        = max(int(request.args.get("page",  1)), 1)
    limit       = min(int(request.args.get("limit", 20)), 100)
    skip        = (page - 1) * limit

    # ── Build filter ───────────────────────────────────────────────────────────
    query = {}
    if team:        query["team"]        = team
    if environment: query["environment"] = environment
    if status:      query["status"]      = status.upper()

    # ── Execute ────────────────────────────────────────────────────────────────
    total  = db.policy_runs.count_documents(query)
    cursor = db.policy_runs.find(query).sort("timestamp", -1).skip(skip).limit(limit)
    runs   = [_serialize(doc) for doc in cursor]

    return jsonify({
        "runs":  runs,
        "total": total,
        "page":  page,
        "limit": limit,
    }), 200


@audit_bp.route("/api/audit/<run_id>", methods=["GET"])
def get_audit_log(run_id):
    from app import db
    if db is None:
        return jsonify({"error": "No database connection"}), 500

    doc = db.policy_runs.find_one({"run_id": run_id})
    if not doc:
        return jsonify({"error": "Run not found"}), 404

    return jsonify(_serialize(doc)), 200
